lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+		);
+
+		// Add new classes.
+		foreach ( $this->modifications->class_changes as $class => $should_include ) {
+			if ( $should_include ) {
+				$new_class .= " {$class}";
+			}
+		}
+
+		$this->set_attribute( 'class', trim( $new_class ) );
+		$this->modifications->class_changes = array();
+	}
+
+
+	public function apply() {
+		$document = $this->scanner->document;
+
+		$this->commit_class_changes();
+
+		/*
+		 * Since we could be asking to modify attributes in a different
+		 * order than where they appear in the HTMl stream we need to
+		 * sort the modification replacements so we can progress
+		 * linearly and in a single pass.
+		 */
+		usort( $this->modifications->replacements, static function ( $a, $b ) {
+			if ( $a[0] === $b[0] ) {
+				return 0;
+			}
+
+			return $a[0] < $b[0] ? -1 : 1;
+		} );
+
+		$output = '';
+		$at = 0;
+		foreach ( $this->modifications->replacements as $replacement ) {
+			list( $start, $end, $text ) = $replacement;
+
+			if ( $start > $at ) {
+				$output .= substr( $document, $at, $start - $at );
+			}
+
+			$output .= $text;
+			$at = $end;
+		}
+
+		$output .= substr( $document, $at );
+		return $output;
+	}
+
+
+	public function __toString() {
+		return $this->apply();
+	}
+}
+
+
+/**
+ * Describes the search conditions for finding a given tag in an HTML document.
+ */
+class WP_Tag_Find_Descriptor {
+	/**
+	 * We're looking for an HTML tag of this name, up to the comparable
+	 * equivalence of those names (lower-cased, Unicode-normalized, etc...).
+	 * If we're looking for "any tag" then this property will be `null`.
+	 *
+	 * h1...h6 are special since they are variations of the same base tag.
+	 * To find "any heading tag" pass the special value `h`.
+	 *
+	 * @var string|null
+	 */
+	public $tag_name;
+
+
+	/**
+	 * We're looking for a tag also containing this CSS class name, up to
+	 * the comparable equivalence of those names. If we're not looking for
+	 * a class name this property will be `null`.
+	 *
+	 * @var string|null
+	 */
+	public $class_name;
+
+
+	/**
+	 * We're looking for a tag that also has some data-attribute values.
+	 * If all we care about is the presence of one of those values this
+	 * will contain the name of that attribute, up to the comparable
+	 * equivalence of that name. If we also want to constrain the value
+	 * of that attribute we will have a pair containing the name of the
+	 * attribute (to equivalence) with a predicate function. If we're
+	 * not looking for a data-attribute this property will be `null`.
+	 *
+	 * Example:
+	 * <code>
+	 *     // has a `data-wp-container` attribute
+	 *     'data-wp-container'
+	 *
+	 *     // has a `data-wp-container` attribute with the value `core/group`
+	 *     ['data-wp-block', 'core/group']
+	 *
+	 *     // has a `data-test-group` attribute with an even-valued id
+	 *     ['data-test-group', function ( $group ) { return 0 === (int) $group % 2; }]
+	 * </code>
+	 *
+	 * @var array<string|callable>
+	 */
+	public $data_attribute;
```

Comment:
I'm not sure if this is an actual use-case. We may be able to remove the data attribute matching – let's keep an eye on it.

Reply:
I think that's right. part of me still thinks it might be more common than we anticipate, but we should try and do some discovery and see how common this is in practice with existing plugins, if we can

Reply:
In general, I'm vouching for adding features later instead of removing them – it's easier, especially with WordPress BC policy. A discovery is a good idea 👍

Reply:
yeah that's a good plan. I'll remove references to it.

eventually I thought it could make sense if we have to go beyond tag name and class simply to add a function for the visitor. every day this looks more and more like a traditional visitor pattern and we could simply allow for that idiom. if we have to scan the entire document up until a termination we wouldn't be losing much by calling a callback on every tag anyway

```php
function render_my_block( $html, $block ) {
    return WP_Tag_Visitor( $html, function ( $tag ) use ( $block ) {
        if ( ! $tag->is( 'div' ) ) {
            return;
        }

        $tag->add_class( 'special-image-wrapper' );
        if ( $block['attrs']['isDisabled'] ) {
            $tag->set_attribute( 'disabled', '' );
        }

        return 'done';
    } );
}
```

that's far less declarative but also way more flexible. come to think of it, this could be a way to "parse HTML" without parsing HTML, because eventually we could do something like expose opening and closing tag events, plus depth. I think that's like how `XmlReader` and `XmlWriter` work as a streaming parser. If someone wants to track DOM structure they can do so _in the application code space_ and, for example, only track as much of it as they want. for instance, maybe they only want to find the first `img` inside a `div` - they can do that while ignoring all other document structure 🤔

Reply:
This is great, let's explore that!

Reply:
@dmsnell, the idea you shared with all the opportunities that `WP_Tag_Visitor` could open is excellent. It would give maximum flexibility with minimal changes in the current implementation.